eth/downloader/statesync.go

```diff
@@ -0,0 +1,245 @@
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package downloader
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+type stateSync struct {
+	d        *Downloader
+	db       ethdb.Database
+	sched    *state.StateSync
+	fetching map[string]*stateReq
+	retry    []common.Hash
+	timeout  chan *stateReq
+
+	quit     chan struct{} // for s.cancel()
+	quitOnce sync.Once
+	wg       sync.WaitGroup
+	err      error
+}
+
+type stateReq struct {
+	peer          *peer
+	items         []common.Hash
+	cancelTimeout chan struct{}
+}
+
+type stateSyncStats struct {
+	done    uint64 // number of entries pulled
+	pending uint64 // number of pending entries
+}
+
+// syncState downloads state with the given root hash.
+func syncState(d *Downloader, root common.Hash, db ethdb.Database) error {
+	return spawnStateSync(d, root, db).wait()
+}
+
+// spawnStateSync starts downloading state with the given root hash.
+func spawnStateSync(d *Downloader, root common.Hash, db ethdb.Database) *stateSync {
+	s := &stateSync{
+		d:        d,
+		db:       db,
+		sched:    state.NewStateSync(root, db),
+		fetching: make(map[string]*stateReq),
+		timeout:  make(chan *stateReq),
+		quit:     make(chan struct{}),
+	}
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		s.err = s.run(d.stateCh)
+	}()
+	return s
+}
+
+// cancel aborts a running state sync. It can be called multiple times.
+func (s *stateSync) cancel() {
+	s.quitOnce.Do(func() {
+		close(s.quit)
+		s.wait()
+	})
+}
+
+// wait blocks until the sync is done or canceled.
+func (s *stateSync) wait() error {
+	s.wg.Wait()
+	return s.err
+}
+
+func (s *stateSync) run(deliver <-chan dataPack) error {
+	defer func() {
+		for _, req := range s.fetching {
+			s.requestDone(req)
+		}
+	}()
+
+	newPeer := make(chan *peer, 200)
+	peerSub := s.d.peers.SubscribeNewPeers(newPeer)
+	defer peerSub.Unsubscribe()
+
+	for {
+		if s.sched.Pending() == 0 {
+			return nil
+		}
+		s.assignTasks()
+
+		select {
+		case packI := <-deliver:
+			pack := packI.(*statePack)
+			req := s.fetching[pack.peerId]
+			if req == nil {
+				log.Debug("Unrequested node data", "peer", pack.peerId, "len", len(pack.states))
+				continue
+			}
+			s.requestDone(req)
+			req.peer.SetNodeDataIdle(pack.Items())
+			if pack.Items() == 0 {
+				log.Trace("Empty node data pack", "peer", pack.peerId)
+				for _, item := range req.items {
+					req.peer.MarkLacking(item)
+				}
+				continue
+			}
+			procStart := time.Now()
+			n, err := s.process(req, pack)
+			if err != nil {
+				log.Warn("Node data write error", "err", err)
+			}
+			s.updateStats(n, time.Since(procStart))
+
+		case req := <-s.timeout:
```

Comment:
This select construct has the same issue as the current code:

```go
select {
  case packI := <-deliver:
    // Process
  case req := <-s.timeout:
}
```

Process can take a long time. By that time all timeouts will fire (even if all the nodes actually delivered too). The end result is that this select statement picks deliver/timeouts randomly if both are available, so it will consider requests timed out that are actually queued up on the deliver channel.